framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js

```diff
@@ -42,47 +43,199 @@ class BlockSynchronizationMechanism {
 	}
 
 	async run(receivedBlock) {
+		this.active = true;
 		try {
-			this.active = true;
 			const { connectedPeers: peers } = await this.channel.invoke(
-				'network:getUniqueOutboundConnectedPeers',
+				'network:getNetworkStatus',
 			);
 			if (!peers.length) {
 				throw new Error('Connected peers list is empty');
 			}
-			// eslint-disable-next-line no-unused-vars
+
 			const bestPeer = this._computeBestPeer(peers);
-			// TODO: handle bestPeer and move on to step 2 defined in
-			// https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md#block-synchronization-mechanism
-			// ...
+			await this._requestAndValidateLastBlock(receivedBlock, bestPeer);
+			await this._revertToLastCommonBlock(receivedBlock, bestPeer);
+		} finally {
+			this.active = false;
+		}
+	}
 
-			// The node requests the last common block C from P (Peer).
-			const { data: networkLastBlock } = await this.channel.invoke(
-				'network:requestFromPeer',
-				{
-					procedure: 'getLastBlock',
-					peerId: bestPeer.id,
-				},
+	/**
+	 * Helper function that encapsulates:
+	 * 1. applying a penalty to a peer.
+	 * 2. restarting sync.
+	 * 3. throwing the reason.
+	 * @param {object} peer - The object that contains the peer ID to target
+	 * @param receivedBlock
+	 * @param {Error } error - An error object containing the reason for applying
+	 * a penalty and restarting sync
+	 * @private
+	 */
+	_applyPenaltyAndRestartSync(peer, receivedBlock, error) {
+		this.channel.invoke('network:applyPenalty', {
+			peerId: peer.id,
+			penalty: 100,
+		});
+		this.channel.publish('chain:processor:sync', { block: receivedBlock });
+		throw error;
+	}
+
+	/**
+	 * Reverts the current chain so the new tip of the chain corresponds to the
+	 * last common block.
+	 * @param receivedBlock
+	 * @param {Object} peer - The selected peer to target.
+	 * @return {Promise<void>}
+	 * @private
+	 */
+	async _revertToLastCommonBlock(receivedBlock, peer) {
+		const lastCommonBlock = await this._requestLastCommonBlock(peer);
+
+		if (!lastCommonBlock) {
+			// TODO: Should we apply penalty here?
+			this._applyPenaltyAndRestartSync(
```

Comment:
In case we don't find a common block from the peer that just send us the conflicting block, we can be sure something went wrong or he modified his database? Check with science if we should apply a penalty and how we should handle this edge case.

Reply:
Solved after talking with Jan